src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -163,162 +173,598 @@ public long GetNumberOfCountersToProcess(DocumentsOperationContext context, stri
+                newEntry->Etag = newETag;
+
+                using (var builder = new ManualBlittableJsonDocumentBuilder<UnmanagedWriteBuffer>(context))
+                {
+                    builder.Reset(BlittableJsonDocumentBuilder.UsageMode.None);
+                    builder.StartWriteObjectDocument();
+
+                    builder.StartWriteObject();
+
+                    builder.WritePropertyName(DbIds);
+
+                    builder.StartWriteArray();
+                    builder.WriteValue(context.Environment.Base64Id);
+                    builder.WriteArrayEnd();
+
+                    builder.WritePropertyName(Values);
+                    builder.StartWriteObject();
+
+                    builder.WritePropertyName(name);
+                    builder.WriteRawBlob(newVal.Ptr, newVal.Length);
+
+                    builder.WriteObjectEnd();
+
+                    builder.WriteObjectEnd();
+                    builder.FinalizeDocument();
+
+                    data = builder.CreateReader();
+                }
+            }
+
+            return data;
         }
 
-        private void PutCounterImpl(DocumentsOperationContext context, string documentId, string collection, string name, string changeVector, long value)
+        public void PutCounters(DocumentsOperationContext context, string documentId, string collection, string changeVector,
+            BlittableJsonReaderObject sourceData)
         {
             if (context.Transaction == null)
             {
                 DocumentPutAction.ThrowRequiresTransaction();
-                Debug.Assert(false);// never hit
+                return;
             }
 
-            var collectionName = _documentsStorage.ExtractCollectionName(context, collection);
-            var table = GetCountersTable(context.Transaction.InnerTransaction, collectionName);
-
-            using (GetCounterKey(context, documentId, name, changeVector ?? context.Environment.Base64Id, out var counterKey))
+            try
             {
-                using (DocumentIdWorker.GetStringPreserveCase(context, name, out Slice nameSlice))
-                using (table.Allocate(out TableValueBuilder tvb))
+                var collectionName = _documentsStorage.ExtractCollectionName(context, collection);
+                var table = GetCountersTable(context.Transaction.InnerTransaction, collectionName);
+
+                BlittableJsonReaderObject data = null;
+                sourceData.TryGet(Values, out BlittableJsonReaderObject sourceCounters);
+
+                using (DocumentIdWorker.GetLowerIdSliceAndStorageKey(context, documentId, out Slice counterKey, out _))
                 {
-                    if (changeVector != null)
+                    using (table.Allocate(out TableValueBuilder tvb))
                     {
-                        if (table.ReadByKey(counterKey, out var existing))
+                        if (changeVector != null && table.ReadByKey(counterKey, out var existing))
```

Comment:
```suggestion
                        if ( table.ReadByKey(counterKey, out var existing))
```

Can't use the CV here

Reply:
fixed